04/app.js

```diff
@@ -4,86 +4,118 @@ import { createRoot } from 'react-dom/client';
-            <section onSubmit={ this.submitHandler }>
-                <form>
-                    <input name="firstName"
-                        value={ firstName }
-                        onChange={ this.inputChange }
-                    />
-                    <input name="lastName"
-                        value={ lastName }
-                        onChange={ this.inputChange }
-                    />
-                    <input type="submit"/>
-                </form>
-                <ul>{ this.renderUsersList() }</ul>
-            </section>
-        );
-    }
-
-    submitHandler = e => {
-        e.preventDefault();
-
-        const { firstName, lastName } = this.state;
-        if(firstName && lastName) {
-            this.addUser(`${firstName} ${lastName}`);
-            this.setState({
-                firstName: '',
-                lastName: '',
-            });
-        } else {
-            // tutaj komunikat dla użytkownika
-        }
-    }
-
-    addUser(name) {
-        this.setState({
-            users: [...this.state.users, name],
-        });
-    }
-
-    removeUser(name) {
-        const currUsers = this.state.users.filter(
-            user => user != name
-        );
-
-        this.setState({
-            users: currUsers,
-        });
-    }
+	state = {
+		firstName: '',
+		lastName: '',
+		searchQuery: '',
+		users: ['Jan Kowalski', 'Michał Nowak'],
+		usersFound: [],
+	};
+
+	renderUsersList() {
+		const { users, searchQuery, usersFound } = this.state;
+
+		if (searchQuery !== '') {
+			return usersFound.map((name) => {
+				return <li>{name}</li>;
+			});
+		} else {
+			return users.map((name) => {
+				return <li onClick={this.clickHandler}>{name}</li>;
+			});
+		}
+	}
+
+	clickHandler = (e) => {
+		const { innerText: userName } = e.target;
+		this.removeUser(userName);
+	};
+
+	inputChange = (e) => {
+		const { name, value } = e.target;
+		this.setState({
+			[name]: value,
+		});
+	};
+
+	render() {
+		const { firstName, lastName, searchQuery } = this.state;
+		return (
+			<>
+				<section onSubmit={this.submitHandler}>
+					<form>
+						<input
+							name='firstName'
+							value={firstName}
+							onChange={this.inputChange}
+						/>
+						<input
+							name='lastName'
+							value={lastName}
+							onChange={this.inputChange}
+						/>
+						<input type='submit' />
+					</form>
+					<ul>{this.renderUsersList()}</ul>
+				</section>
+				<section onChange={this.searchHandler}>
+					<label>Znajdź użytkownika: </label>
+					<input
+						name='searchQuery'
+						value={searchQuery}
+						onChange={this.inputChange}></input>
+				</section>
+			</>
+		);
+	}
+
+	searchHandler = (e) => {
+		const value = e.target.value;
+		this.searchUsers(value);
+	};
+
+	searchUsers(searchQuery) {
+		const { users } = this.state;
+
+		const searchUsers = users.filter((user) => {
+			if (user.includes(searchQuery)) {
+				return user;
+			}
+		});
+
+		this.setState({
+			usersFound: searchUsers,
+		});
```

Comment:
W zasadzie ten element nie jest potrzebny. Wystarczy, aby ta funkcja `return searchUsers`, a jej wywołanie było wew. render i też będzie ok. Wtedy tylko w jednym miejscu przechowujemy użytkowników. To o tyle będzie lepsze, że wiemy gdzie są aktualne dane i tylko na tym miejscu się opieramy.